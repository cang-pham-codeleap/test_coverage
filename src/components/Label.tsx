import React from "react";

/**
 * Props for the Label component
 * @property {string} htmlFor - The ID of the form element this label is associated with
 * @property {React.ReactNode} children - The content to be displayed within the label
 * @property {string} [className] - Additional CSS classes to apply to the label
 * @property {'default' | 'error' | 'required'} [variant='default'] - The visual style of the label
 * @property {boolean} [srOnly=false] - Whether the label should be visible only to screen readers
 * @property {React.HTMLAttributes<HTMLLabelElement>} [props] - Additional HTML attributes to apply to the label
 */
export interface LabelProps
  extends React.LabelHTMLAttributes<HTMLLabelElement> {
  htmlFor: string;
  children: React.ReactNode;
  className?: string;
  variant?: "default" | "error" | "required";
  srOnly?: boolean;
}

/**
 * Label component for form elements
 *
 * @example
 * ```tsx
 * <Label htmlFor="email">Email Address</Label>
 * <Label htmlFor="password" variant="required">Password</Label>
 * <Label htmlFor="terms" srOnly>Accept Terms and Conditions</Label>
 * ```
 */
export const Label: React.FC<LabelProps> = ({
  htmlFor,
  children,
  className = "",
  variant = "default",
  srOnly = false,
  ...props
}) => {
  // Determine variant-specific styles
  const variantClasses = {
    default: "text-gray-700",
    error: "text-red-500",
    required:
      'text-gray-700 after:content-["*"] after:ml-0.5 after:text-red-500',
  };

  // Create CSS classes based on props
  const classes = [
    "text-sm font-medium",
    variantClasses[variant],
    srOnly ? "sr-only" : "",
    className,
  ]
    .filter(Boolean)
    .join(" ");

  return (
    <label htmlFor={htmlFor} className={classes} {...props}>
      {children}
    </label>
  );
};

export default Label;
