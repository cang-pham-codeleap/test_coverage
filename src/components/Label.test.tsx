/**
 * Label component tests
 * - Tests basic rendering
 * - Tests different variants
 * - Tests screen reader only mode
 * - Tests custom class application
 */

import React from "react";
import { render, screen } from "@testing-library/react";
import Label from "./Label";

describe("Label Component", () => {
  // Test basic rendering
  test("renders label with children", () => {
    render(<Label htmlFor="test-input">Test Label</Label>);
    const labelElement = screen.getByText("Test Label");
    expect(labelElement).toBeInTheDocument();
    expect(labelElement).toMatchSnapshot();
  });

  // Test default variant
  test("renders default variant by default", () => {
    render(<Label htmlFor="test-input">Default Label</Label>);
    const labelElement = screen.getByText("Default Label");
    expect(labelElement.className).toContain("text-gray-700");
  });

  // Test error variant
  test("renders error variant when specified", () => {
    render(<Label htmlFor="test-input" variant="error">Error Label</Label>);
    const labelElement = screen.getByText("Error Label");
    expect(labelElement.className).toContain("text-red-500");
  });

  // Test required variant
  test("renders required variant with asterisk styling", () => {
    render(<Label htmlFor="test-input" variant="required">Required Field</Label>);
    const labelElement = screen.getByText("Required Field");
    expect(labelElement.className).toContain("after:content-[\"*\"]");
    expect(labelElement.className).toContain("after:text-red-500");
  });

  // Test screen reader only mode
  test("applies sr-only class when srOnly is true", () => {
    render(<Label htmlFor="test-input" srOnly>Screen Reader Only Label</Label>);
    const labelElement = screen.getByText("Screen Reader Only Label");
    expect(labelElement.className).toContain("sr-only");
  });

  // Test htmlFor attribute
  test("applies htmlFor attribute correctly", () => {
    render(<Label htmlFor="test-id">Label with For Attribute</Label>);
    const labelElement = screen.getByText("Label with For Attribute");
    expect(labelElement).toHaveAttribute("for", "test-id");
  });

  // Test custom class name
  test("applies custom className when provided", () => {
    const customClass = "my-custom-class";
    render(<Label htmlFor="test-input" className={customClass}>Custom Style</Label>);
    const labelElement = screen.getByText("Custom Style");
    expect(labelElement.className).toContain(customClass);
  });

  // Test additional attributes
  test("applies additional HTML attributes", () => {
    render(<Label htmlFor="test-input" data-testid="custom-label">Custom Attribute</Label>);
    const labelElement = screen.getByTestId("custom-label");
    expect(labelElement).toMatchSnapshot();
  });
});
