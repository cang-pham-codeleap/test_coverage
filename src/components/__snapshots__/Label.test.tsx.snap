// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Component applies additional HTML attributes 1`] = `
<label
  class="text-sm font-medium text-gray-700"
  data-testid="custom-label"
  for="test-input"
>
  Custom Attribute
</label>
`;

exports[`Label Component renders label with children 1`] = `
<label
  class="text-sm font-medium text-gray-700"
  for="test-input"
>
  Test Label
</label>
`;
